refactor: Move many functions into LegacyScriptPubKeyMan and further separate it from CWallet

achow101 opened the pull request:
Moves several more key management and metadata functions into LegacyScriptPubKeyMan from CWallet to further separate the two.

Note to reviewers: All of the `if (auto spk_man = walletInstance->m_spk_man.get()) {` blocks will be replaced with for loops in the next PR so you may see some things in those blocks that don't necessarily make sense with an `if` but will with a `for`.

achow101 committed:
Refactor: Move RewriteDB code out of CWallet

This commit does not change behavior.

achow101 committed:
Refactor: Move SetupGeneration code out of CWallet

This commit does not change behavior.

achow101 committed:
Refactor: Move nTimeFirstKey accesses out of CWallet

This commit does not change behavior.

achow101 committed:
Refactor: Move MarkUnusedAddresses code out of CWallet::AddToWalletIfInvolvingMe

This commit does not change behavior.

achow101 committed:
Refactor: Move SetWalletFlag out of LegacyScriptPubKeyMan::UpgradeKeyMetadata

This commit does not change behavior.

achow101 committed:
Refactor: Declare LegacyScriptPubKeyMan methods as virtual

This commit does not change behavior.

achow101 committed:
Refactor: Move SetAddressBookWithDB call out of LegacyScriptPubKeyMan::ImportScriptPubKeys

This commit does not change behavior.

achow101 committed:
Refactor: Move SetAddressBook call out of LegacyScriptPubKeyMan::GetNewDestination

This commit does not change behavior.

achow101 committed:
Refactor: Remove UnsetWalletFlag call from LegacyScriptPubKeyMan::SetHDSeed

This commit does not change behavior.

achow101 committed:
Refactor: Move GetKeypoolSize code out of CWallet

This commit does not change behavior.

achow101 committed:
Refactor: Move Upgrade code out of CWallet::CreateWalletFromFile

This commit does not change behavior.

achow101 committed:
Refactor: Move LoadKey LegacyScriptPubKeyMan method definition

This commit does not change behavior.

achow101 committed:
Refactor: Move GetMetadata code out of getaddressinfo

Easier to review ignoring whitespace:

    git log -p -n1 -w

This commit does not change behavior.

achow101 committed:
Refactor: Add new ScriptPubKeyMan virtual methods

This commit does not change behavior.

achow101 committed:
MOVEONLY: Reorder LegacyScriptPubKeyMan methods

Can verify move-only with:

    git log -p -n1 --color-moved

This commit is move-only and doesn't change code or affect behavior.

achow101 committed:
Refactor: Move HavePrivateKeys code out of CWallet::CreateWalletFromFile

This commit does not change behavior.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -3623,15 +3625,12 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(interfaces::Chain& chain,
-        }
-
-        // Top up the keypool
-        if (walletInstance->m_spk_man->CanGenerateKeys() && !walletInstance->m_spk_man->TopUp()) {
```
In commit "Refactor: Move SetupGeneration code out of CWallet" (f95750d52a21a69ad55db8422f10c563fb9e08ac)

Note for other reviews: Previous code was calling TopUp, while new code is calling SetupGeneration which calls NewKeyPool which calls TopUpKeyPool. I think the effect is basically the same because fFirstRun is true here so the additional ErasePool calls in NewKeyPool should have no effect.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -210,9 +210,11 @@ WalletCreationStatus CreateWallet(interfaces::Chain& chain, const SecureString&
-            wallet->m_spk_man->NewKeyPool();
+            {
+                if (auto spk_man = wallet->m_spk_man.get()) {
+                    spk_man->SetupGeneration();
```
In commit "Refactor: Move SetupGeneration code out of CWallet" (f95750d52a21a69ad55db8422f10c563fb9e08ac)

Should this be checking the return code from SetupGeneration and setting an error if it fails? Old code wasn't checking the return value from NewKeyPool either, but that doesn't seem right.

achow101 replied:
It should. Done.

ryanofsky reviewed (approved):
Code review ACK e1b1702449a9172185d4bfb5e9043748328132ee. Left comments and suggestions which could be followed up later if we want to avoid losing momentum on #17261

achow101 committed:
Remove SetWalletFlag from WalletStorage

SetWalletFlag is unused.

Does not change any behavior

achow101 committed:
refactor: Replace UnsetWalletFlagWithDB with UnsetBlankWalletFlag in ScriptPubKeyMan

ScriptPubKeyMan is only using UnsetWalletFlagWithDB to unset the blank
wallet flag. Just make that it's own function and not expose the flag
writing directly.

This does not change behavior.

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3756,26 +3756,24 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
-        if (it != pwallet->m_script_metadata.end()) {
-            meta = &it->second;
+        if (!meta) {
+            meta = spk_man->GetMetadata(CScriptID(scriptPubKey));
```
From the ScriptPubKeyMan design view, it doesn't make sense to have different metadata for keys and scripts, at least exposed publicly to the wallet. It should really be `GetMetadata` that takes a scriptPubKey, but it fit better to use uint160 for right now.

achow101 commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -320,8 +320,7 @@ void LegacyScriptPubKeyMan::UpgradeKeyMetadata()
     }
-    batch.reset(); //write before setting the flag
-    m_storage.SetWalletFlag(WALLET_FLAG_KEY_ORIGIN_METADATA);
+    batch.reset();
```
Done

achow101 commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -320,8 +320,7 @@ void LegacyScriptPubKeyMan::UpgradeKeyMetadata()
         }
     }
-    batch.reset(); //write before setting the flag
-    m_storage.SetWalletFlag(WALLET_FLAG_KEY_ORIGIN_METADATA);
```
Added a commit dropping SetWalletFlag. Also checked it isn't being used in Descriptor wallets.

achow101 commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -874,7 +874,8 @@ void LegacyScriptPubKeyMan::SetHDSeed(const CPubKey& seed)
     NotifyCanGetAddressesChanged();
-    m_wallet.UnsetWalletFlag(WALLET_FLAG_BLANK_WALLET);
+    WalletBatch batch(m_storage.GetDatabase());
+    m_storage.UnsetWalletFlagWithDB(batch, WALLET_FLAG_BLANK_WALLET);
```
Added a commit changing those to UnsetBlankWalletFlag. Also checked DescriptorScriptPubKeyMan doesn't use UnsetWalletFlag.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -1401,9 +1401,17 @@ bool CWallet::ImportScriptPubKeys(const std::string& label, const std::set<CScri
+    if (!spk_man->ImportScriptPubKeys(script_pub_keys, have_solving_data, timestamp)) {
         return false;
     }
+    WalletBatch batch(*database);
```
Done

achow101 commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -362,6 +362,41 @@ bool LegacyScriptPubKeyMan::CanGetAddresses(bool internal)
 
+bool LegacyScriptPubKeyMan::Upgrade(int prev_version, std::string& error)
+{
+    AssertLockHeld(cs_wallet);
```
Removed

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/17304#discussion_r340314846

> Removed

The lock assert probably needs to be added back. I'm not sure why clang for me wasn't showing the same error, but clang on travis legitimately errors that it needs the assert to call `MarkPreSplitKeys():

```
wallet/scriptpubkeyman.cpp:399:9: error: calling function 'MarkPreSplitKeys' requires holding mutex 'cs_wallet' exclusively [-Werror,-Wthread-safety-analysis]
        MarkPreSplitKeys();
```

achow101 replied:
Added it back in.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -3794,8 +3794,12 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(interfaces::Chain& chain,
+            time_first_key = std::min(time_first_key, spk_man->GetTimeFirstKey());
+        }
+        if (time_first_key) {
+            if (Optional<int> first_block = locked_chain->findFirstBlockWithTimeAndHeight(time_first_key - TIMESTAMP_WINDOW, rescan_height, nullptr)) {
```
Made it an optional

ryanofsky reviewed (approved):
Code review ACK 5631c01c845b1fb9a91e0b40709ceaac204fbb5d. Just a bunch of suggested changes made since the last review.

ryanofsky commented:
Not sure what your preference is @achow101, but it might be a good idea to replace #17261 with #17304 on the high priority review list https://github.com/bitcoin/bitcoin/projects/8, since #17261 builds on #17304.

This PR is also just making small and mostly obvious changes, so it should be more approachable for reviewers.

achow101 commented:
> Not sure what your preference is @achow101, but it might be a good idea to replace #17261 with #17304 on the high priority review list https://github.com/bitcoin/bitcoin/projects/8, since #17261 builds on #17304.

Yeah, it should replace that. #17261 was added to the list before I made this pr.

ryanofsky reviewed (approved):
Code review ACK b04a526aea91ba4caddbde1a1313a1dbf46bc5e5. Only change is adding back missing lock assert.

ryanofsky commented:
This PR has 18 commits, so it could be useful to get some Approach ACKS here on whether it's a reasonable size, or should be made smaller by dropping some number of commits, or dropping particular commits that seem more difficult to review. All of the commits should be small and straightforward, but it is also easily possible to scale back this PR if necessary.

instagibbs commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -265,6 +265,31 @@ bool LegacyScriptPubKeyMan::EncryptKeys(CKeyingMaterial& vMasterKeyIn)
 
+bool LegacyScriptPubKeyMan::GetReservedDestination(const OutputType type, bool internal, int64_t& index, CKeyPool& keypool)
+{
+    {
```
why the scope here?

achow101 replied:
A lock is going to be added in the next step

ryanofsky replied:
> A lock is going to be added in the next step

The scope isn't needed even with the lock and all other changes from #17261:

https://github.com/bitcoin/bitcoin/blob/35b0f7b5447bce4eef360c7e4237a928bb0eba3c/src/wallet/scriptpubkeyman.cpp#L275-L290

But I don't think it's a big deal. Could clean this up later if reservekey and reservedestination methods are merged later (https://github.com/bitcoin/bitcoin/pull/17304#discussion_r340311958)

achow101 replied:
I'll just leave this as is so it can get merged sooner :)

instagibbs commented on src/wallet/wallet.cpp:
```diff
@@ -3804,8 +3804,13 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(interfaces::Chain& chain,
+        Optional<int64_t> time_first_key;
+        if (auto spk_man = walletInstance->m_spk_man.get()) {
+            int64_t time = spk_man->GetTimeFirstKey();
+            if (!time_first_key || time < *time_first_key) time_first_key = time;
```
I realize this is stopping mid-way, but this was slightly confusing. Add a comment saying this is temporarily going to always be true until a loop is introduced?

achow101 replied:
That generally applies throughout this PR. So I guess a note to reviewers in the thread should be good enough?

instagibbs replied:
ok! Consider this your note, future reviewers

instagibbs reviewed (approved):
ACK with non-blocking nits

"Refactor: Move GetMetadata code out of getaddressinfo" more accurately: "Refactor: Remove direct accesses to key metadata code from getaddressinfo"

same type of message nit for "Refactor: Move HavePrivateKeys code out of CWallet::CreateWalletFromFile" and "Refactor: Move SetupGeneration code out of CWallet"

Sjors commented:
I'm going through these commits now. I've seen this code a few times before, but if others want to eject one or more commits to get it merged faster, that could make sense.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -561,11 +566,11 @@ bool CWallet::EncryptWallet(const SecureString& strWalletPassphrase)
-            m_spk_man->SetHDSeed(m_spk_man->GenerateNewSeed());
+        if (auto spk_man = m_spk_man.get()) {
+            if (spk_man->IsHDEnabled()) {
+                spk_man->SetupGeneration(true);
```
In commit "Refactor: Move SetupGeneration code out of CWallet" (ef97f0b5451e4959c0d0bf649a71eec9b226210e)

No need to change now, but it seems like it'd probably be a good idea to check for an error from SetupGeneration here as well.

Sjors commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -287,6 +287,23 @@ bool LegacyScriptPubKeyMan::TopUp(unsigned int size)
+            WalletLogPrintf("%s: Detected a used keypool key, mark all keypool key up to this key as used\n", __func__);
+            MarkReserveKeysAsUsed(mi->second);
+
+            if (!TopUpKeyPool()) {
```
I'm also a bit confused about the distinction between `TopUpKeyPool()` and `Topup()`. In the native descriptor wallet PR there's an internal call to `Topup()` inside `DescriptorScriptPubKeyMan::GetNewDestination` and `DescriptorScriptPubKeyMan::MarkUnusedAddresses`: https://github.com/bitcoin/bitcoin/pull/16528/files#diff-5462ceb8a760a507152ab8b76bd48774R1459

achow101 replied:
In `LegacyScriptPubKeyMan`, `TopUp()` and `TopUpKeyPool()` are the same. Since `TopUpKeyPool` was already being used throughout existing `CWallet` code that was moved into `LegacyScriptPubKeyMan`, I decided to let those calls just keep using `TopUpKeyPool`. But for the `ScriptPubKeyMan` interface itself, I wanted it to be named `TopUp` because that conceptually made more sense than calling it `TopUpKeyPool`.

ryanofsky replied:
Any reason you would not want to rename TopUpKeyPool method to TopUp and call it everywhere instead of having two LegacyScriptPubKeyMan methods that do the same thing?

Sjors commented:
Code review ACK up to 6f34d7450be333fbcd3616d2d4305bf38b528ad3 (`Move MarkUnusedAddresses code out of CWallet::AddToWalletIfInvolvingMe`).  Will review the remaining 5 commits later.

promag commented on src/wallet/wallet.cpp:
```diff
@@ -1406,9 +1406,19 @@ bool CWallet::ImportScriptPubKeys(const std::string& label, const std::set<CScri
         return false;
     }
+    if (apply_label) {
+        WalletBatch batch(*database);
```
26fef3a525fd98f8205fd2f859f476fbf2cc6fcd

Just noting that there's actually a behavior change here - now if `LegacyScriptPubKeyMan::ImportScriptPubKeys` fails __no label is applied__.

Also, another change, which I think is harmless, is that now 2 batches are always used - easily "reverted" by passing the batch to `LegacyScriptPubKeyMan::ImportScriptPubKeys` from this method.

achow101 replied:
Those are harmless. `ImportScriptPubKeys` only fails when writing to the database fails, so you'll have bigger problems if that happens.

promag reviewed (commented):
Light code review ACK b04a526aea91ba4caddbde1a1313a1dbf46bc5e5 with some comments.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -253,6 +253,7 @@ void CWallet::UpgradeKeyMetadata()
     if (m_spk_man) {
         m_spk_man->UpgradeKeyMetadata();
     }
+    SetWalletFlag(WALLET_FLAG_KEY_ORIGIN_METADATA);
```
re: https://github.com/bitcoin/bitcoin/pull/17304#discussion_r341791687

> [2590266](https://github.com/bitcoin/bitcoin/commit/2590266c193d27c979ad6287d72555ca3df630bc)
> 
> What if `IsLocked()`?

Nice catch! This seems like a bug. I think either the `m_storage.IsLocked() || m_storage.IsWalletFlagSet(WALLET_FLAG_KEY_ORIGIN_METADATA)` check should move from `LegacyScriptPubKeyMan::UpgradeKeyMetadata` to
`CWallet::UpgradeKeyMetadata`, or preferably, the`SetWalletFlag(WALLET_FLAG_KEY_ORIGIN_METADATA)` call should move back in the other direction.

It's also unclear why `UpgradeKeyMetadata` shoud be a virtual method instead of being specific to the LegacyScriptPubKeyMan class. It doesn't seem like non-legacy keyman types should implement this upgrade function or be accessing this wallet flag.

ryanofsky replied:
It seems like the `CWallet::UpgradeKeyMetadata()` and `ScriptPubKeyMan::::UpgradeKeyMetadata()` methods could both be dropped and `walletdb.cpp` could call `pwallet->GetLegacyScriptPubKeyMan()` and `LegacyScriptPubKeyMan::UpgradeKeyMetadata` directly for compatibility.

 to do here would be to drop the

achow101 replied:
Changed to be non-virtual and moved that check out to CWallet.

achow101 commented on src/wallet/scriptpubkeyman.h:
```diff
@@ -282,6 +296,7 @@ class LegacyScriptPubKeyMan : public ScriptPubKeyMan, public FillableSigningProv
     //! Fetches a pubkey from mapWatchKeys if it exists there
     bool GetWatchPubKey(const CKeyID &address, CPubKey &pubkey_out) const;
 
+    /* SigningProvider overrides */
```
Is what necessary?

promag replied:
The category comment.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -3038,12 +3038,17 @@ bool CWallet::TopUpKeyPool(unsigned int kpSize)
-        result = spk_man->GetNewDestination(type, label, dest, error);
+        result = spk_man->GetNewDestination(type, dest, error);
+    }
+    if (result) {
```
Does it really matter?

promag replied:
No, just that it looks like a leftover where before result could be changed in between. 

Actually I think this would look better without the result variable.

You can resolve this.

ryanofsky reviewed (approved):
Code review ACK 707a8178109b75fb672f233f57c3c5db952ad5f1. Only change since last review is the fix to UpgradeKeyMetaData to avoid setting WALLET_FLAG_KEY_ORIGIN_METADATA when the wallet is locked.

laanwj commented:
code review ACK 152b0a00d8e681dd098f6b548447b82ab54ebe3c

Sjors commented:
re-ACK 152b0a00d8e681dd098f6b548447b82ab54ebe3c

If you can append a commit to get rid of `unused variable` warnings in rpcdump.cpp that would be nice though (see #17338 by @jonatack).

```
 wallet/rpcdump.cpp:137:28: error: unused variable 'spk_man' [-Werror,-Wunused-variable]
    LegacyScriptPubKeyMan& spk_man = GetLegacyScriptPubKeyMan(*wallet);
```

promag commented:
Code review ACK 152b0a00d8e681dd098f6b548447b82ab54ebe3c.
